[Job Launcher] GCV integration

eugenvoronov opened the pull request:
## Issue tracking
The progress of this issue is tracked in [Job Launcher] GCV integration
[#2905](https://github.com/humanprotocol/human-protocol/issues/2905).

## Context behind the change
This pull request integrates Google Cloud Vision (GCV) to perform content moderation. The key changes include:
- Using GCV's capabilities to detect and moderate images for inappropriate content.
- Preventing job launches and fund refunds for datasets with a "LIKELY" or "VERY_LIKELY" rating for inappropriate content.
- Save images urls to google cloud storate for datasets containing "POSSIBLE" ratings and sending Slack notifications.
- Skipping the content moderation step for whitelisted users while ensuring their datasets are still annotated.
- Implemented new `JobModeration` service.
- Added new migration `1738052876684-updateJobStatusAndCronJobType`.
   - Added new cron_job types: `job-moderation` and `parse-job-moderation-results`
   - Added new jobs statuses: `on_moderation`, `moderation_passed` and `possible_abuse_in_review`
- Implemented `slack` util.
- Implemented `sleep` util (Method was removed, because it is no longer used.).
- Implemented new methods to Google Cloud Storage utils and unit tests
   - isGCSBucketUrl - Validates if a given URL is a valid Google Cloud Storage URL.
   - convertToGCSPath - Converts a valid Google Cloud Storage HTTP URL to a GCS path.
   - convertToHttpUrl - Converts a GCS path to a valid Google Cloud Storage HTTP URL.
- Added new env variables: 
```
# Vision
GOOGLE_PROJECT_ID=
GOOGLE_PRIVATE_KEY=
GOOGLE_CLIENT_EMAIL=
GCV_MODERATION_RESULTS_BUCKET=
GCV_MODERATION_RESULTS_FILES_PATH=

# Slack
SLACK_ABUSE_NOTIFICATION_WEBHOOK_URL=
```
- Implemented `asyncBatchAnnotateImages`
- Implemented `batchAnnotateImages`  (Logic were removed, because it is no longer used.).

**Limitations:**
`batchAnnotateImages` can process up to 16 images per request and has a rate limit of 1800 requests per minute.
`asyncBatchAnnotateImages` can handle up to 2000 images in a single request, which is more suitable for processing large batches of images. However, the requirement for images to be hosted in Google Cloud Storage makes it less versatile for our use case.

## How has this been tested?
The following testing steps were undertaken:
- Unit tests were updated and new tests added to cover the combined moderation and annotation process.
- Manual testing with various datasets, including those from whitelisted users, was conducted to verify the correct functioning of all changes.

## Release plan
To release this change:
1. Update migrations, for compatibility, data from the cron jobs table should be removed.
2. Update environment variables to include the necessary GCV credentials.
3. Create new bucket in Google Cloud Storage with private access for **positive abuse and possible** results.
4. Deploy the updated service, ensuring GCV permissions and access are correctly configured.
5. Monitor the first few job runs closely for any unexpected behaviors or errors.

portuu3 commented on packages/apps/job-launcher/server/src/modules/cron-job/cron-job.service.spec.ts:
```diff
@@ -55,6 +55,7 @@ import { ethers } from 'ethers';
 import { NetworkConfigService } from '../../common/config/network-config.service';
 import { QualificationService } from '../qualification/qualification.service';
 import { WhitelistService } from '../whitelist/whitelist.service';
+import { JobModerationService } from '../job/job-moderation.service';
```
There no tests for the moderation cron job

portuu3 commented on packages/apps/job-launcher/server/src/modules/job/job-moderation.service.ts:
```diff
@@ -0,0 +1,289 @@
+    const dataModerationResults: DataModerationResultDto =
+      await this.dataModeration(manifest.data.data_url);
+
+    if (dataModerationResults.containsAbuse) {
```
This property is actually unncessary, you can check `veryLikelyOrLikelyResults` length. 
Actually you can use a different name for `veryLikelyOrLikelyResults`, like `positiveAbuseResults`

portuu3 commented on packages/apps/job-launcher/server/src/modules/job/job-moderation.service.ts:
```diff
@@ -0,0 +1,289 @@
+      jobEntity.status = JobStatus.FAILED;
+      await this.jobRepository.updateOne(jobEntity);
+
+      return jobEntity;
```
Let's send a message to slack also for positive abuse.

portuu3 commented on packages/apps/job-launcher/server/src/modules/job/job-moderation.service.ts:
```diff
@@ -0,0 +1,289 @@
+    }
+
+    // If there are possible moderation issues, save links to GCS and send a Slack notification
+    if (dataModerationResults.possibleResults.length > 0) {
```
Just to follow the same line of `positiveAbuseResults`, rename it to `possibleAbuseResults`

portuu3 commented on packages/apps/job-launcher/server/src/modules/job/job-moderation.service.ts:
```diff
@@ -0,0 +1,289 @@
+      await this.dataModeration(manifest.data.data_url);
+
+    if (dataModerationResults.containsAbuse) {
+      const abusiveImageLinks = dataModerationResults.veryLikelyOrLikelyResults
```
Since this is reused in possible abuse links, create a function for this.
We should upload a file with the abusive content as we do with possible abuse.
The only difference between possible and confirmed abuse is the status set to the job.

portuu3 commented on packages/apps/job-launcher/server/src/modules/job/job-moderation.service.ts:
```diff
@@ -0,0 +1,289 @@
+      );
+    }
+
+    jobEntity.status = JobStatus.MODERATION_PASSED;
```
For possible abuse, the status should not be set to MODERATION_PASSED.
We should probably use a new status POSSIBLE_ABUSE_IN_REVIEW

portuu3 commented on packages/apps/job-launcher/server/src/modules/job/job-moderation.service.ts:
```diff
@@ -0,0 +1,477 @@
+      const allResults: protos.google.cloud.vision.v1.IAnnotateImageResponse[] =
+        [];
+
+      for (let i = 0; i < imageUrls.length; i += batchSize) {
```
Lets remove sync code

eugenvoronov replied:
yes, you are right, a more complex implementation will require saving the results of each batch, and in case of an error, continuing from the next batch.

dnechay reviewed (commented):
I didn't tested it locally and suppose we should give it a try on staging first before we merge

eugenvoronov commented on packages/apps/job-launcher/server/src/modules/job/job-moderation.service.ts:
```diff
@@ -0,0 +1,477 @@
+    }
+  }
+
+  public async asyncDataModeration(dataUrl: string): Promise<void> {
```
I've tested it with 550 images, I will test it with +2000

portuu3 replied:
As per documentation it should not process more than 2000 by request. What I mean by this comment is that we are not handling this scenario.

eugenvoronov replied:
In Progress

eugenvoronov commented on packages/apps/job-launcher/server/src/common/utils/storage.ts:
```diff
@@ -127,3 +127,78 @@ export async function listObjectsInBucket(url: URL): Promise<string[]> {
+ * @returns {boolean} - Returns true if the URL is valid, otherwise returns false.
+ */
+export function isGCSBucketUrl(url: string): boolean {
+  const gcsHttpRegex =
```
I think so, if the user passes an invalid string that only contains `http://.storage.googleapis.com/`?

dnechay replied:
In that case it might be better to split this validation in multiple steps to easily digest the code:
1. Check if it's valid URL in general (the should be something likse `isValidUrl` already in the code)
2. Then you check if it's GCS url
3. Then you check if it has bucket name in its domain name or in url path

There are some cases that we can't handle, like if they have GCS on custom domain or can provide valid GCS url, but miss last character in it 🤷🏻 , so IMU we either:
- have simplest possible validation, i.e. make sure it's GCS domain
- or we have rich validation, e.g. keep regex but cover it with excessive amount of unit-tests, making it hard to break the regex accidentally

E.g., here are some URLs that pass this validation:
- anything where bucket names is less than 3 symbols, e.g. `https://oh.storage.googleapis.com` (GCS doesn't allows <3 chars bucket names)
- `https://----.storage.googleapis.com`
- `https://test.storage.googleapis.com/\` - it won't pass `new URL` constructor that might be used internally by Minio
- `https://test.storage.googleapis.com/with    whitespaces`
- `https://test-.storage.googleapis.com`
- `https://-test.storage.googleapis.com`

And also some URLs that can't pass it, when should, e.g. `https://my_bucket.storage.googleapis.com`

eugenvoronov commented on packages/apps/job-launcher/server/src/common/utils/storage.ts:
```diff
@@ -127,3 +127,78 @@ export async function listObjectsInBucket(url: URL): Promise<string[]> {
+export function isGCSBucketUrl(url: string): boolean {
+  const gcsHttpRegex =
+    /^https:\/\/([a-zA-Z0-9\-]+)\.storage\.googleapis\.com(?:\/(.*))?$/;
+  const gcsGsRegex = /^gs:\/\/[a-zA-Z0-9\-]+(\/.*)?$/;
```
I think we should check the whole URL against the pattern to make sure the URL is valid.

dnechay replied:
In case with `gs` it's not actually URL, it's `gs://` protocol, then `bucket_name`, then path, so bucket name should also include `_` (maybe some other checks that correspond to valid GS bucket name) and also make sure that path doesn't contain invalid characters

https://cloud.google.com/storage/docs/buckets#naming

eugenvoronov commented on packages/apps/job-launcher/server/src/common/utils/storage.ts:
```diff
@@ -127,3 +127,78 @@ export async function listObjectsInBucket(url: URL): Promise<string[]> {
+ */
+export function isGCSBucketUrl(url: string): boolean {
+  const gcsHttpRegex =
+    /^https:\/\/([a-zA-Z0-9\-]+)\.storage\.googleapis\.com(?:\/(.*))?$/;
```
Yes, when user create a job, we parse StorageDataDto to this format `bucketName.storage.googleapis.com` using `generateBucketUrl` function on Storage Module.

eugenvoronov commented on packages/apps/job-launcher/server/src/modules/cron-job/cron-job.service.ts:
```diff
@@ -79,6 +82,86 @@ export class CronJobService {
+          await this.jobModerationService.jobModeration(jobEntity);
+        } catch (err) {
+          const errorId = uuidv4();
+          const failedReason = `${ErrorJobModeration.InappropriateContent} (Error ID: ${errorId})`;
```
Thanks, I changed it
In fact this error tells where the failure occurred, and in the logs we can see the real reason.

eugenvoronov commented on packages/apps/job-launcher/server/src/modules/job/job-moderation.service.ts:
```diff
@@ -0,0 +1,477 @@
+      const dataFiles = await listObjectsInBucket(new URL(dataUrl + '/'));
+
+      const gcDataUrl = convertToGCSPath(dataUrl);
+      const validFiles = dataFiles.filter(
```
Yes, for example, there may be such a case when in the directory with the dataset that the user sent `dataUrl`, another directory is stored together with his images and in order to filter it, I suggested using a filter.

eugenvoronov commented on packages/apps/job-launcher/server/src/modules/job/job-moderation.service.ts:
```diff
@@ -0,0 +1,477 @@
+    isConfirmedAbuse: boolean,
+  ): Promise<void> {
+    const bucketName = isConfirmedAbuse
+      ? this.visionConfigService.moderationResultsBucket
```
This is a bug, I reworked the logic and did not change this method. Thanks!

eugenvoronov commented:
within the GCV PR code review I fixed:
- Fixed dependencies 
- Fixed UNDER_MODERATION job status and updated migrations
- Fully removed sync batch annotation logic 
- Removed sleep util, because its no longer used
- Improved sendSlackNotification method
- Implemented new GCSUtil
- Existing methods GCSUtil starting to be more generic
- Implemented constructGcsPath to GCSUtil 
- Implemented new unit tests for GCSUtil
- Updated existing unit tests 
- And some some small changes

portuu3 commented on packages/apps/job-launcher/server/src/modules/content-moderation/gcv-content-moderation.service.ts:
```diff
@@ -66,6 +67,12 @@ export class GCVContentModerationService implements IContentModeratorService {
+
+    // Initialize cache with expiration time of 1 hour and check period of 30 minutes
+    this.bucketListCache = new NodeCache({
+      stdTTL: 60 * 60,
```
I would use ttl 30 min, a GCV check should not take more than that
